### PR TITLE
[core] Add shared object ref to effects

### PR DESCRIPTION
**Pull request by gdanezis**
Add the object ref of shared object inputs to the transaction effects, addressing #1580 .

### CODE REVIEW

File: sui_core/src/authority.rs [hunk 1]
```diff
@@ -329,10 +330,13 @@ impl AuthorityState {
 
         // At this point we need to check if any shared objects need locks,
         // and whether they have them.
-        let has_shared_object = objects_by_kind
+        let shared_object: Vec<_> = objects_by_kind
```
Comment (asonnino) [inline sui_core/src/authority.rs:333]: Should this be called `shared_objects` since we may have many of them?

File: sui_core/src/authority.rs [hunk 2]
```diff
@@ -355,6 +359,7 @@ impl AuthorityState {
             transaction_digest,
         );
         let effects = execution_engine::execute_transaction_to_effects(
+            shared_object,
```
Comment (asonnino) [inline sui_core/src/authority.rs:362]: So we explicitly pass this argument to avoid re-calling `.compute_object_reference()` within the executing engine?

**Review by asonnino — APPROVED** [review]

File: sui_core/src/authority.rs (hunk 2 above)
Comment (gdanezis) [reply sui_core/src/authority.rs:362]: Yeah basically -- computing digests is not for free. So this leads to computing once, and allocating once.

File: sui_core/src/authority.rs (hunk 1 above)
Comment (gdanezis) [reply sui_core/src/authority.rs:333]: Yes, obviously!

File: sui_core/src/authority.rs (hunk 2 above)
Comment (asonnino) [reply sui_core/src/authority.rs:362]: cool just making sure I didn't miss anything important

File: sui_core/src/authority.rs [hunk 3]
```diff
@@ -329,10 +330,13 @@ impl AuthorityState {
...
-            .any(|(kind, _)| matches!(kind, InputObjectKind::SharedMoveObject(_)));
-        if has_shared_object {
+            .filter(|(kind, _)| matches!(kind, InputObjectKind::SharedMoveObject(_)))
+            .map(|(_, obj)| obj.compute_object_reference())
+            .sorted()
+            .collect();
+        if !shared_objects.is_empty() {
             // If the transaction contains shared objects, we need to ensure they have been scheduled
             // for processing by the consensus protocol.
             self.check_shared_locks(&transaction_digest, &objects_by_kind)
```
Comment (lxfind) [inline sui_core/src/authority.rs:342]: Might as well pass in `shared_objects`

File: sui_core/src/authority.rs [hunk 4]
```diff
@@ -329,10 +330,13 @@ impl AuthorityState {
 
         // At this point we need to check if any shared objects need locks,
         // and whether they have them.
-        let has_shared_object = objects_by_kind
+        let shared_objects: Vec<_> = objects_by_kind
```
Comment (lxfind) [inline sui_core/src/authority.rs:333]: nit: may be clearer to rename all `shared_objects` to `shared_object_refs`

**Review by lxfind — APPROVED** [review]
Thank you!
Some nit comments

File: sui_core/src/authority.rs (hunk 4 above)
Comment (gdanezis) [reply sui_core/src/authority.rs]: Good idea, did that!

**Pull request merged**